src/js/trove/matrix.js

```diff
@@ -0,0 +1,1310 @@
+            return makeMatrix(self.$h, c, new_arr);
+        }
+
+        //Internal function : Swaps rows m,n in place
+        var swapRows = function (self, m, n) {
+            arity(3, arguments, "lup-mat", false);
+            for (i = 0; i < self.$w; i++) {
+                temp = get1dElem(self, m, i);
+                self.$underlyingMat[get1dpos(m, i, self.$w)] = get1dElem(self, n, i);
+                self.$underlyingMat[get1dpos(n, i, self.$w)] = temp;
+            }
+        }
+        /* Calculates LUP decomposition PA = LU where:
+         P is permutation matrix
+         L is lower triangular
+         U is upper triangular 
+         return tuple of (P,L,U,exchanges) where exchanges is the no of row swaps*/
+        var LUPMat = function (self) {
+            arity(1, arguments, "lup-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "lup-mat", self, annMatrix);
+            if (!isSquareMatrix(self)) {
+                return runtime.ffi.throwMessageException("LU decomposition with partial pivoting requires square matrix");
+            }
+            N = self.$h;
+            P = IdentityMatrix(N);
+            A = duplicateMatrix(self);
+            L = IdentityMatrix(N);
+            U = duplicateMatrix(self);
+            exchanges = runtime.makeNumber(0);
+
+            let getMaxEntry = function (cIdex) {
+
+                let maxIdex = cIdex
+                max = runtime.num_abs(get1dElem(A, cIdex, cIdex));
+                let dim = A.$h;
+
+
+                for (i = cIdex + 1; i < dim; i++) {
+
+                    let next = runtime.num_abs(get1dElem(A, i, cIdex));
+
+                    if (next > max) {
+                        max = next;
+                        maxIdex = i;
+                    }
+
+                }
+
+                return maxIdex;
+            };
+
+            let pivot = function (p, n) {
+                let dim = A.$h;
+                let temp, i;
+                // U
+                swapRows(U, p, n);
+
+                // L
+                swapRows(L, p, n);
+                swapRows(P, p, n);
+                exchanges = runtime.plus(exchanges, 1);
+            };
+
+            let eliminate = function (p) {
+                let dim = A.$h;
+                let i, j;
+
+                for (i = p + 1; i < dim; i++) {
+                    let l = runtime.divide(get1dElem(U, i, p), get1dElem(U, p, p));
+                    L.$underlyingMat[get1dpos(i, p, dim)] = l;
+                    for (j = p; j < dim; j++) {
+                        U.$underlyingMat[get1dpos(i, j, dim)] = runtime.minus(get1dElem(U, i, j), runtime.times(l, get1dElem(U, p, j)));
+                    }
+                }
+
+            };
+            let PLU = function () {
+
+                let dim = A.$h;
+                let i, j, k;
+                for (i = 0; i < dim - 1; i++) {
+                    // Find the max entry
+                    let maxIdex = getMaxEntry(i);
```

Comment:
Did you really save much typing or readability by leaving out the `n` in `maxIndex`?